Prevent assigning PR authors to their own PRs

marcos-arranz opened the pull request:
This PR fixes codeassigner assigning the author as a reviewer, and also a possible infinite loop.

andersfischernielsen commented on src/assign.ts:
```diff
@@ -187,6 +187,11 @@ export const selectReviewers = async (
       }
       debug(`Found random team member: ${randomTeamMember}.`)
 
+      if (randomTeamMember && randomTeamMember === author) {
+        debug(`'${randomTeamMember}' is the author '${randomTeamMember}'. Skipping.`)
+        continue
+      }
```
We already assert that `randomTeamMember` isn't `undefined | null` in L:181 😄 

```suggestion
      if (randomTeamMember === author) {
        debug(`'${randomTeamMember}' is the author '${author}'. Skipping.`)
        continue
      }
```

andersfischernielsen commented on src/assign.ts:
```diff
@@ -153,7 +152,8 @@ export const selectReviewers = async (
     const randomGlobalCodeowners = randomize(globalCodeowners)
-    const selected = randomFileOwner ?? randomGlobalCodeowner(randomGlobalCodeowners)
+    const selected = randomFileOwner ?? randomGlobalCodeowners?.[0]
+    if (selected && !isTeam(selected)) randomGlobalCodeowners?.shift()
```
`selected` being `undefined | null` is already handled in the following two checks and we already `break` if `!selected`, so this check is redundant 😄 

```suggestion
    if (!isTeam(selected)) randomGlobalCodeowners?.shift()
```

andersfischernielsen replied:
I'd rather we change `isTeam` to handle a `string?` parameter 😄 to keep the logic consistent

andersfischernielsen commented on src/assign.test.ts:
```diff
@@ -524,6 +524,51 @@ describe('Reviewer selection', () => {
+    expect(result.users.every(name => name !== options.author)).toBeTruthy()
+  })
+
+  it('does not select an author from CODEOWNERS when owner is a team', async () => {
```
```suggestion
  it('does not select individual authors from CODEOWNERS when owners are teams', async () => {
```

andersfischernielsen commented on src/assign.test.ts:
```diff
@@ -524,6 +524,51 @@ describe('Reviewer selection', () => {
     expect(result.users.every(name => name !== options.author)).toBeTruthy()
   })
 
+  it('does not select an author from CODEOWNERS with assignIndividuals true', async () => {
```
Doesn't this break the behavior of the Action? 

If I have a repository where the `CODEOWNERS` looks like this:

```
* @some-org/cool-team
```

where the team consists of the members `@some-username` and I set `assignIndividuals: true`

then I'd expect the result of the Action running to be assigning `@some-username` to the PR and NOT the global team name. It doesn't look like that's the case anymore 🤔

marcos-arranz replied:
No, that's should not be the case. The behaviour to select an individual or not hasn't been changes. I have added a test case with that example to cover it.

andersfischernielsen commented:
I'm a bit scared we're 
- breaking the expected functionality of the Action
- re-introducing logic that's been problematic in the past wrt. the infinite looping - I spent a decent chunk of time avoiding that exact loop a while back

with this 😄

marcos-arranz commented:
> I'm a bit scared we're
* breaking the expected functionality of the Action
* re-introducing logic that's been problematic in the past wrt. the infinite looping - I spent a decent chunk of time avoiding that exact loop a while back
with this 😄

Thanks for your comments @andersfischernielsen, I have addressed most of them. Regarding your fears, I have added a test case to cover the first point. Regarding the second case, could you point me which code can fall into a infinite loop? I have added some code to prevent, because there were some conditions were it surface it as in this test run https://github.com/pleo-io/assign-random-codeowners/actions/runs/4146429489/jobs/7172152599

andersfischernielsen reviewed (approved):
LGTM! 🍰
